internal/tar/tar.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2021 Tetrate
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tar
+
+import (
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/mholt/archiver/v3"
+	"github.com/ulikunitz/xz"
+)
+
+// NewDecompressor returns a compression function based on the "src" filename.
+// NOTE: As of May 2021, all Envoy binaries except the first are tar.xz
+func NewDecompressor(src string, r io.Reader) (io.Reader, error) {
+	if strings.HasSuffix(src, "tar.xz") {
```

Comment:
I think it's more hermetic if we use `xz.ValidHeader()` to check if the file is xz-compressed by checking binary header

Reply:
hmm if we do that we need to pass a buffered reader and peek. it is a little more complicated vs just letting it fail later..

Reply:
I suppose you are interested in getting rid of the need to read the src name? It could make sense if inlined.

Reply:
yeah because the file name isn't actually a way of declaring the file type.. but this is a subtle change so you could ignore my suggestion 😄

Reply:
if you don't mind, let me handle polishing this in #216 after merge?

Reply:
sg!